File: split/flushnodelist.pas
procedure flushnodelist(p: halfword);
label
  30;
var
  q: halfword;
begin
  while p <> 0 do
  begin
    q := mem[p].hh.rh;
    if (p >= himemmin) then
    begin
      mem[p].hh.rh := avail;
      avail := p;
      {dynused:=dynused-1;}

    end
    else
    begin
      case mem[p].hh.b0 of
        0, 1, 13:
        begin
          flushnodelist(mem[p + 5].hh.rh);
          freenode(p, 7);
          goto 30;
        end;
        2:
        begin
          freenode(p, 4);
          goto 30;
        end;
        3:
        begin
          flushnodelist(mem[p + 4].hh.lh);
          deleteglueref(mem[p + 4].hh.rh);
          freenode(p, 5);
          goto 30;
        end;
        8:{1358:}

        begin
          case mem[p].hh.b1 of
            0: freenode(p, 3);
            1, 3:
            begin
              deletetokenref(mem[p + 1].hh.rh);
              freenode(p, 2);
              goto 30;
            end;
            2, 4: freenode(p, 2);
            others: confusion(1294)
          end;
          goto 30;
        end{:1358}
;
        10:
        begin
          begin
            if mem[mem[p + 1].hh.lh].hh.rh = 0 then
              freenode(mem[p + 1].hh.lh, 4)
            else
              mem[mem[p + 1].hh.lh].hh.rh := mem[mem[p + 1].hh.lh].hh.rh - 1;
          end;
          if mem[p + 1].hh.rh <> 0 then
            flushnodelist(mem[p + 1].hh.rh);
        end;
        11, 9, 12: ;
        6: flushnodelist(mem[p + 1].hh.rh);
        4: deletetokenref(mem[p + 1].int);
        7:
        begin
          flushnodelist(mem[p + 1].hh.lh);
          flushnodelist(mem[p + 1].hh.rh);
        end;
        5: flushnodelist(mem[p + 1].int);
        {698:}
14:
        begin
          freenode(p, 3);
          goto 30;
        end;
        15:
        begin
          flushnodelist(mem[p + 1].hh.lh);
          flushnodelist(mem[p + 1].hh.rh);
          flushnodelist(mem[p + 2].hh.lh);
          flushnodelist(mem[p + 2].hh.rh);
          freenode(p, 3);
          goto 30;
        end;
        16, 17, 18, 19, 20, 21, 22, 23, 24, 27, 26, 29, 28:
        begin
          if mem[p + 1].hh.rh >= 2 then
            flushnodelist(mem[p + 1].hh.lh);
          if mem[p + 2].hh.rh >= 2 then
            flushnodelist(mem[p + 2].hh.lh);
          if mem[p + 3].hh.rh >= 2 then
            flushnodelist(mem[p + 3].hh.lh);
          if mem[p].hh.b0 = 24 then
            freenode(p, 5)
          else if mem[p].hh.b0 = 28 then
            freenode(p, 5)
          else
            freenode(p, 4);
          goto 30;
        end;
        30, 31:
        begin
          freenode(p, 4);
          goto 30;
        end;
        25:
        begin
          flushnodelist(mem[p + 2].hh.lh);
          flushnodelist(mem[p + 3].hh.lh);
          freenode(p, 6);
          goto 30;
        end;
        {:698}
others: confusion(353)
      end;
      freenode(p, 2);
      30: ;
    end;
    p := q;
  end;
end;

File: cpp/src/flushnodelist.cpp
#include "flushnodelist.h"
#include "freenode.h"
#include "deleteglueref.h"
#include "lecture.h"
#include "erreur.h"

void flushnodelist(halfword p)
{
	while (p)
	{
		auto q = link(p);
		if (p >= himemmin)
		{
			link(p) = avail;
			avail = p;
		}
		else
		{
			switch (type(p))
			{
				case hlist_node:
				case vlist_node:
				case unset_node:
					flushnodelist(list_ptr(p));
					freenode(p, box_node_size);
					break;
				case rule_node:
					freenode(p, rule_node_size);
					break;
				case ins_node:
					flushnodelist(ins_ptr(p));
					deleteglueref(ins_ptr(p));
					freenode(p, ins_node_size);
					break;
				case whatsit_node:
					switch (subtype(p))
					{
						case open_node:
							freenode(p, open_node_size);
							break;
						case write_node:
						case special_node:
							deletetokenref(write_tokens(p));
							freenode(p, write_node_size);
							break;
						case close_node:
						case language_node: 
							freenode(p, small_node_size);
							break;
						default: 
							confusion("ext3");
					}
					break;
				case glue_node:
					deleteglueref(glue_ptr(p));
					if (leader_ptr(p))
						flushnodelist(leader_ptr(p));
					freenode(p, 2);
					break;
				case kern_node:
				case math_node:
				case penalty_node: 
					freenode(p, 2);
					break;
				case ligature_node: 
					flushnodelist(lig_ptr(p));
					freenode(p, 2);
					break;
				case mark_node: 
					deletetokenref(mark_ptr(p));
					freenode(p, 2);
					break;
				case disc_node:
					flushnodelist(pre_break(p));
					flushnodelist(post_break(p));
					freenode(p, 2);
					break;
				case adjust_node: 
					flushnodelist(adjust_ptr(p));
					freenode(p, 2);
					break;
				case style_node:
					freenode(p, style_node_size);
					break;
				case choice_node:
					flushnodelist(display_mlist(p));
					flushnodelist(text_mlist(p));
					flushnodelist(script_mlist(p));
					flushnodelist(script_script_mlist(p));
					freenode(p, style_node_size);
					break;
				case ord_noad:
				case op_noad:
				case bin_noad:
				case rel_noad:
				case open_noad:
				case close_noad:
				case punct_noad:
				case inner_noad:
				case radical_noad:
				case over_noad:
				case under_noad:
				case vcenter_noad:
				case accent_noad:
					if (math_type(nucleus(p)) >= sub_box)
						flushnodelist(nucleus(p));
					if (math_type(supscr(p)) >= sub_box)
						flushnodelist(info(supscr(p)));
					if (math_type(subscr(p)) >= sub_box)
						flushnodelist(info(subscr(p)));
					if (type(p) == radical_noad || type(p) == accent_noad)
						freenode(p, radical_noad_size);
					else
						freenode(p, noad_size);
					break;
				case left_noad:
				case right_noad:
					freenode(p, noad_size);
					break;
				case fraction_noad:
					flushnodelist(info(numerator(p)));
					flushnodelist(info(denominator(p)));
					freenode(p, fraction_noad_size);
					break;
				default: 
					confusion("flushing"); 
			}
		}
		p = q;
	}
}
